Provide async mode for local autograd engine.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31230 Provide async mode for local autograd engine.**

A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

pritamdamania committed:
Provide async mode for local autograd engine.

A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]

pritamdamania87 commented:
cc @jjlilley @aazzolini

kostmo commented:
## CircleCI build failures summary
As of commit dce49c3:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=dce49c33eac95691c208b8d77ae2fd6c92be2588).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3942842) pytorch_linux_xenial_py2_7_9_test (1/1)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=73799389))
```
Dec 13 03:30:41 Ran 28 tests in 5.449s 
Dec 13 03:30:41  
Dec 13 03:30:41 FAILED (failures=1, skipped=3) 
Dec 13 03:30:41  
Dec 13 03:30:41 Generating XML reports... 
Dec 13 03:30:41 Traceback (most recent call last): 
Dec 13 03:30:41   File "test/run_test.py", line 456, in <module> 
Dec 13 03:30:41     main() 
Dec 13 03:30:41   File "test/run_test.py", line 449, in main 
Dec 13 03:30:41     raise RuntimeError(message) 
Dec 13 03:30:41 RuntimeError: test_utils failed! 
Dec 13 03:30:41 + cleanup 
Dec 13 03:30:41 + retcode=1 
Dec 13 03:30:41 + set +x 
Dec 13 03:30:41 =================== sccache compilation log =================== 
Dec 13 03:30:41 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Dec 13 03:30:41  
Dec 13 03:30:41 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 13 03:30:41 Compile requests                109 
Dec 13 03:30:41 Compile requests executed        27 
Dec 13 03:30:41 Cache hits                       20 
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

jjlilley commented:
fwiw: my initial read here is that the code looks quite reasonable - I tried going through corner-cases, though I'm not an expert in this module.

Given that this is for a real pain point and the diff is incremental in nature (i.e. it doesn't actually make things async yet), we should probably try to get this in, barring objections.
Not sure if others see red flags here?

(Regardless, Phabricator side has a minor merge issue)

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -808,6 +765,45 @@ variable_list Engine::execute_with_graph_task(
+
+  // Not a worker
+  if (worker_device == NO_DEVICE) {
+    return graph_task->future_;
```
This would be skipping the post-processing in this case?

jjlilley replied:
It looks like this is intentional, right? The post-processing is supposed to be done by the entity fulfilling the future (i.e. mark_graph_task_completed), rather than after the future is fulfilled.

Semi-related, I am mildly questioning why we're holding the graph_task->mutex_ lock across post-processing (if nothing else, it might be nice to drop the lock right before calling graph_task->future_->markCompleted(vval)), but this doesn't necessarily look like an actual lock scope change vs. previousl.

pritamdamania87 replied:
@albanD As @jjlilley mentioned this is done as part of `mark_graph_task_completed`. Regarding the locking, I just kept the same behavior as we already had. Before this PR all the post processing was run under the `graph_task->mutex_`.

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -313,45 +319,30 @@ auto Engine::thread_main(
-  // oscillating between the two threads.
-  if (reentrant_thread && current_depth == 0) {
-    graph_task->not_done_.notify_all();
+    auto base_owner = local_graph_task->owner_;
```
Why don't we need to notify the owner thread when it is not a worker anymore? Is it because it just returns a Future now and the `mark_graph_task_completed` above will notify on the future?
If it is the case, a comment would be useful.

pritamdamania87 replied:
Yes, the future takes care of this now. I will add a comment.

albanD reviewed (changes requested):
Looks quite good.
I am still a bit confused of when is the post-processing is called depending on the different cases. Adding a comment at the beginning of the file with the different cases will help.

ezyang commented:
Adding @bddppq and @wanchaol as they are concurrently working on  autograd engine changes related on where cpu thread executes

ezyang commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -103,7 +103,7 @@ void DistAutogradContext::addOutstandingRpc(
           // If we have an error, let the local autograd engine know about it.
           std::runtime_error err((*futErr).what());
-          graphTask_->set_exception(std::make_exception_ptr(err), nullptr);
+          graphTask_->set_exception(err, nullptr);
```
Hmm? What's going on here?

pritamdamania87 replied:
The generic Future uses `setError` to mark exceptions and there is no way to retrieve an error from exception_ptr. As a result, I ensured we instead always pass a reference to the exception.

ezyang commented on torch/csrc/autograd/engine.h:
```diff
@@ -26,6 +27,8 @@ struct ReadyQueue;
 
 namespace torch { namespace autograd {
 
+using FutureVariableList = torch::utils::Future<variable_list>;
```
nit: I personally find the rhs clearer than the type alias.

pritamdamania87 replied:
This would actually look like this in the signature: `std::shared_ptr<torch::utils::Future<variable_list>>`. I felt it was more compact writing just `std::shared_ptr<FutureVariableList>`.

gchanan replied:
In speech, I'd also definitely refer to this as a "VariableListFuture" over "FutureVariableList", because "VariableList" tells me a lot more about the type than Future.

ezyang commented:
For my code reviewing sanity could you split this into a stack diff: first diff move the functions from header to cpp file, second diff actually make the changes?

pritamdamania87 commented on torch/csrc/utils/future.h:
```diff
@@ -9,7 +9,7 @@ namespace utils {
 // std::string error message
 class TORCH_API FutureError final : public std::exception {
-public:
+ public:
```
Yes.

pritamdamania87 commented:
> I am still a bit confused of when is the post-processing is called depending on the different cases. Adding a comment at the beginning of the file with the different cases will help.

I added comments for these inline.

> For my code reviewing sanity could you split this into a stack diff: first diff move the functions from header to cpp file, second diff actually make the changes?

Not sure I followed this completely, which functions should I move from header to cpp?

pritamdamania committed:
Update on "Provide async mode for local autograd engine."


A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Provide async mode for local autograd engine."


A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -313,45 +319,33 @@ auto Engine::thread_main(
-  // a thread which is at its maximum stack depth and they keep exiting right
-  // after. We will always switch to a new thread for each call, so, we'll keep
-  // oscillating between the two threads.
-  if (reentrant_thread && current_depth == 0) {
```
seems like we don't handle this case in the new future way anymore. we should mark the future as completed for this condition as well to make sure parent thread know it.

pritamdamania87 replied:
IIUC, wouldn't each invocation of `thread_main` at least enter the while loop once (since outstanding tasks would be > 0)? If so, wouldn't `mark_graph_task_completed` inside the loop notify the future appropriately? Are there any cases where this wouldn't happen?

wanchaol replied:
hmm I think you are right. If it is a re-entrant thread, then the loop will only terminate when `outstanding_tasks_ == 0`, and that will trigger the future being marked as complete, and should be ok for us to remove this.

wanchaol commented on torch/csrc/autograd/engine.h:
```diff
@@ -96,13 +95,17 @@ struct GraphTask {
 
+  // Future representing the completion of the graph task. Notified when all
+  // tasks are done.
+  std::shared_ptr<FutureVariableList> future_;
```
can we call this something like `future_result_`? it's hard to tell what's holding inside this future from the var name.

pritamdamania committed:
Update on "Provide async mode for local autograd engine."


A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]

ezyang commented:
> Not sure I followed this completely, which functions should I move from header to cpp?

It seemed like in your diff you moved some functions around: specifically `graph_task_completed`, `execute_with_graph_task` (I guess you didn't actually move these functions to C++ file, they stayed in the header). If you can rearrange the functions so that these functions stay in the same location, that would greatly help the readability of the diff.

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -158,6 +158,12 @@ int NodeTask::getReentrantDepth() const {
 }
 
+bool graph_task_completed(const std::shared_ptr<GraphTask>& graph_task) {
+  std::unique_lock<std::mutex> lock(graph_task->mutex_);
```
I don't really understand why you're taking out the lock now, when you didn't previously, because it seems like you didn't change the body of this function and it's still accessing atomics. What's going on? Do you still need the atomics now?

pritamdamania87 replied:
You are correct, I don't need the lock here. I think I added it for extra safety but on closer inspection we don't really need the lock.

pritamdamania87 reviewed (commented):
> It seemed like in your diff you moved some functions around: specifically graph_task_completed, execute_with_graph_task (I guess you didn't actually move these functions to C++ file, they stayed in the header). If you can rearrange the functions so that these functions stay in the same location, that would greatly help the readability of the diff.

Had to move `graph_task_completed` due to the dependency from `thread_main`. Will try to keep `execute_with_graph_task` in the same place.

pritamdamania committed:
Update on "Provide async mode for local autograd engine."


A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Provide async mode for local autograd engine."


A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]

pritamdamania87 commented:
@albanD @ezyang Could you take another look? Thanks for the feedback so far!

pritamdamania committed:
Update on "Provide async mode for local autograd engine."


A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]

albanD reviewed (changes requested):
Looks pretty good. All the comments really help to follow the code.
Small logic issues with `execute_with_graph_task`.

pritamdamania87 commented on test/test_utils.py:
```diff
@@ -122,7 +122,8 @@ def test_checkpoint_valid(self):
         modules = list(model.children())
         out = checkpoint_sequential(modules, chunks, input_var)
-        with self.assertRaisesRegex(RuntimeError, "Checkpointing is not compatible"):
+        # python_error in case of py2_7_9.
```
To be honest I couldn't figure out why python 2 didn't report appropriate exceptions whereas python 3 did. Given that python 2 is going away very soon, I didn't spend too much time here.

albanD replied:
Is there a way to mark this so that it will be easily identified and removed when we remove python 2?

pritamdamania87 replied:
Do we have any existing conventions to handle something like this? I can create a gh issue with the 'better-engineering' tag to remove this once python2 goes away.

albanD replied:
Or maybe python 2 is already away? 🎉 
But if you want to merge before that, a gh issue is good I think.

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -770,11 +753,30 @@ variable_list Engine::execute_with_graph_task(
+
+void Engine::mark_graph_task_completed(std::shared_ptr<GraphTask>& graph_task) {
+  std::unique_lock<std::mutex> lock(graph_task->mutex_);
+  if (graph_task->future_result_->completed()) {
```
It could be possible if we `setError` as part of `GraphTask::set_exception` an then call `mark_graph_task_completed` from `thread_main`

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -770,11 +753,30 @@ variable_list Engine::execute_with_graph_task(
+    auto val = graph_task_exec_post_processing(graph_task);
+    graph_task->future_result_->markCompleted(val);
+  } catch (std::exception& e) {
+    graph_task->future_result_->setError(e.what());
```
It will notify all the threads waiting on the future and the thread calling wait() will also throw an exception.

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -770,11 +753,30 @@ variable_list Engine::execute_with_graph_task(
+  // Check for errors, call callbacks and sync streams. We return a completed
+  // future here since 'thread_main' above is a call blocking an autograd engine
+  // thread and not the thread the user called 'execute_with_graph_task' from.
+  return std::make_shared<FutureVariableList>(
```
Good point, I'll refactor this once we can confirm the behavior for `total_depth` as I've mentioned above.

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -747,18 +729,19 @@ variable_list Engine::execute_with_graph_task(
-          lock, [&graph_task] { return graph_task_completed(*graph_task); });
+      // graph_task_exec_post_processing is done when the Future is marked as
+      // completed in mark_graph_task_completed.
+      return graph_task->future_result_;
```
Actually, looking into the codebase, is `total_depth` used anywhere? We seem to maintain it across threads but we never use the value?

pritamdamania87 replied:
It is used here to set `reentrant_depth_` in GraphTask: https://github.com/pytorch/pytorch/blob/877c96cddfebee00385307f9e1b1f3b4ec72bfdc/torch/csrc/autograd/engine.cpp#L709. `reentrant_depth_` is used to order items in the priority_queue.

albanD replied:
Ho right. My bad.
current depth is the depth of the current thread while total depth is the depth counting cases where you use the thread pool. SO you do want it for both codepath. If you move it as you suggest, it will just become the same as current depth.

pritamdamania87 replied:
It won't be the same as current depth if I move it, the reason being is that the total_depth is passed into the GraphTask as reentrant_depth and in other threads the total_depth is bootstrapped from reentrant_depth here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L375. Also, note that the incremented value of total_depth is never used in this codepath. The codepath only calls `add_thread_pool_task` which does not use total_depth.

albanD replied:
Right in this codepath, the value stored in the graphtask is used directly.
It is fine to increment it only where we use it I think. So the same place as current_depth.

pritamdamania committed:
Update on "Provide async mode for local autograd engine."


A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]

pritamdamania87 commented:
@albanD Updated the PR based on our discussion of `total_depth`.

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -747,34 +729,58 @@ variable_list Engine::execute_with_graph_task(
+      return graph_task->future_result_;
     } else {
+      // Total depth needs to be updated only in this codepath, since it is
+      // not used in the block above (when we call add_thread_pool_task).
```
nit: maybe mention that `graphtask.reentrant_depth_` is used instead of the TLS as we will change thread in the other branch.

albanD reviewed (approved):
LGTM, thanks!

pritamdamania committed:
Update on "Provide async mode for local autograd engine."


A major issue with distributed autograd currently is that we block an
RPC thread when we call Engine::execute_with_graph_task.

To resolve this issue, I've made modifications to the local autograd engine
such that `execute_with_graph_task` returns a Future instead. The `execute()`
methods for Engine::execute() and DistEngine::execute() still wait() on this
Future which ensures there is no change in behavior yet.

In follow up PRs we can modify the distributed autograd engine to take
advantage of this Future.

Closes #26359

Differential Revision: [D18999709](https://our.internmc.facebook.com/intern/diff/D18999709/)

[ghstack-poisoned]
